sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequencesProcessorDoFn.java

```diff
@@ -0,0 +1,272 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.ordered;
+
+import org.apache.beam.sdk.coders.BooleanCoder;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.extensions.ordered.ProcessingState.ProcessingStateCoder;
+import org.apache.beam.sdk.state.OrderedListState;
+import org.apache.beam.sdk.state.StateSpec;
+import org.apache.beam.sdk.state.StateSpecs;
+import org.apache.beam.sdk.state.TimeDomain;
+import org.apache.beam.sdk.state.Timer;
+import org.apache.beam.sdk.state.TimerSpec;
+import org.apache.beam.sdk.state.TimerSpecs;
+import org.apache.beam.sdk.state.ValueState;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.TupleTag;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Main Stateful DoFn used to process events in the global sequence mode.
+ *
+ * @param <EventT>
+ * @param <EventKeyT>
+ * @param <ResultT>
+ * @param <StateT>
+ */
+class GlobalSequencesProcessorDoFn<
+        EventT, EventKeyT, ResultT, StateT extends MutableState<EventT, ResultT>>
+    extends ProcessorDoFn<EventT, EventKeyT, ResultT, StateT> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GlobalSequencesProcessorDoFn.class);
+
+  private static final String BATCH_EMISSION_TIMER = "batchTimer";
+
+  @TimerId(BATCH_EMISSION_TIMER)
+  @SuppressWarnings("unused")
+  private final TimerSpec batchTimerSpec = TimerSpecs.timer(TimeDomain.EVENT_TIME);
```

Comment:
Hm, noting that I think the usage of this timer will cause issues if the input has allowed lateness. We may need to disallow handling late data here

Reply:
Because this type of timer doesn't honor the lateness?

Reply:
Yeah. I guess we could also look at the input allowed lateness and use that to key the timer though.

Reply:
Good point! Added the logic to capture the lateness of the input and add it to the max timestamp.